os/drivers/input/touch.c

```diff
@@ -0,0 +1,373 @@
+	priv = filep->f_inode->i_private;
+
+	if (!priv) {
+		return -EINVAL;
+	}
+
+	touch_semtake(&priv->sem, false);
+	if (priv->crefs == 0) {
+		priv->dev->touch_enable(priv->dev, true);
+	}
+	priv->crefs++;
+	DEBUGASSERT(priv->crefs > 0);
+	touch_semgive(&priv->sem);
+	return OK;
+}
+
+/****************************************************************************
+ * Name: touch_close
+ ****************************************************************************/
+
+static int touch_close(FAR struct file *filep)
+{
+	FAR struct touch_s *priv;
+	priv = filep->f_inode->i_private;
+
+	if (!priv) {
+		return -EINVAL;
+	}
+
+	touch_semtake(&priv->sem, false);
+	DEBUGASSERT(priv->crefs > 0);
+	priv->crefs--;
+	if (priv->crefs == 0) {
+		priv->dev->touch_enable(priv->dev, false);
+#ifndef CONFIG_DISABLE_POLL
+		/* Check if this file is registered in a list of waiters for polling.
+		* For example, when task A is blocked by calling poll and task B try to terminate task A,
+		* a pollfd of A remains in this list. If it is, it should be cleared.
+		*/
+		(void)touch_semtake(&priv->pollsem, false);
+		for (int i = 0; i < CONFIG_TOUCH_NPOLLWAITERS; i++) {
+			struct pollfd *fds = priv->fds[i];
+			if (fds && (FAR struct file *)fds->filep == filep) {
+				priv->fds[i] = NULL;
+			}
+		}
+		touch_semgive(&priv->pollsem);
+#endif
+	}
+	touch_semgive(&priv->sem);
+	return OK;
+}
+
+/****************************************************************************
+ * Name: touch_read
+ ****************************************************************************/
+
+static ssize_t touch_read(FAR struct file *filep, FAR char *buffer, size_t buflen)
+{
+	FAR struct touch_s *priv;
+	size_t outlen;
+	int ret;
+
+	priv = filep->f_inode->i_private;
+
+	if (!priv || !buffer || buflen < 1) {
+		return -EINVAL;
+	}
+	/* Wait for semaphore to prevent concurrent reads */
+	touch_semtake(&priv->sem, false);
+
+	/* Read the touch data, only if screen has been touched or if we're waiting for touch up */
+	outlen = sizeof(struct touch_sample_s);
+	ret = priv->dev->touch_read(priv->dev, buffer);
+	touch_semgive(&priv->sem);
+	return ret < 0 ? ret : outlen;
+}
+
+/****************************************************************************
+ * Name: touch_write
+ ****************************************************************************/
+
+static ssize_t touch_write(FAR struct file *filep, FAR const char *buffer, size_t buflen)
+{
+	//TODO
+	return OK;
+}
+
+#ifndef CONFIG_DISABLE_POLL
+
+/****************************************************************************
+ * Name: touch_pollnotify
+ ****************************************************************************/
+
+void touch_pollnotify(FAR struct touch_dev_s *priv, pollevent_t eventset)
+{
+	int itr;
+	for (itr = 0; itr < CONFIG_TOUCH_NPOLLWAITERS; itr++) {
+		struct pollfd *fds = g_dev->fds[itr];
+		if (fds) {
+			fds->revents |= (fds->events & eventset);
+			if (fds->revents != 0) {
+				sem_post(fds->sem);
```

Comment:
Could you let me know who (where) waits the sem?

Reply:
Hello Mr. chang,
The application waits for the semaphore to be release:
https://github.com/Samsung/TizenRT/blob/1a92b37ecf206530025b9560e83b43db5053b34a/apps/examples/lcd_test/example_lcd.c#L392C3-L392C16

Reply:
App waits driver's sem post directly?? It looks strange. 
If app is not running after driver sem post due to some thread priority or interrupt, and user touches a screen continiousely, then driver will call sem_post continousely and it will get sem count overflow. Could you check this is possible error or not?